_posts/2015-09-28-service-discovery.md

```diff
@@ -0,0 +1,57 @@
+categories: tech-stories
+authors: ["Dirk Jablonski"]
+---
+
+As epages is heading for a new, microservice-based architecture, new challanges arise. In this kind of architecture, there are a lot of services, which are usually deployed redundantly for high availability. But from a client perspective, all these service instances should be transparent. The client should not need to care about multiple instances of a services, their different locations on the network or if they are operational, down for maintenance or in a failure state. A client should only need to deal with a single address of the service.
+
+The solution to this challange is *Service Discovery*. It is the means by which any service is able to find other services without the need to know about the actual location or other details. In our solution, Service Discovery means that you only need to know a generic URI to communicate with other REST-based services (or even some infrastructure components like message brokers).
+
+#How does it work?
+A service discovery solution usually consists of three core building blocks:
+
+* A *service registry* which holds the data about currently available services and instances
+* A *health-check* mechanism to monitor service health
+* A *lookup* or *routing* mechanism to connect to services
+
+## Service Registry
+At the heart of the service discovery infrastructure, there is a so-called service registry. This registry has the knowledge about all available services and their instances. It gains this knowledge by providing an API to register and deregister hosts or service instances, so the services have a central point to make themselves available to the public.
+
+####Consul
+We chose [Consul](https://consul.io/) as our service registry implementation. Consul provides the core functionality of a service registry by utilizing an agent-based setup. This means, an agent, which is kind of a daemon process, runs on every machine that provides services. Here is a short abstract of Consul's introduction documentation to get to know the most important parts:
+
+>Consul is a distributed, highly available system. [...]
+>Every node that provides services to Consul runs a *Consul agent*.  Running an agent is not required for discovering other services or getting/setting key/value data. The agent is responsible for health checking the services on the node as well as the node itself.
+>The agents talk to one or more *Consul servers*. The Consul servers are where data is stored and replicated. The servers themselves elect a leader.
+>[...]
+>Components of your infrastructure that need to discover other services or nodes can query any of the Consul servers
+>or any of the Consul agents. The agents forward queries to the servers automatically.
+
+For more details, see the full [introduction](https://www.consul.io/intro/index.html)  and the [in-depth architecture overview](https://www.consul.io/docs/internals/architecture.html) on the Consul web site.
```

Comment:
remove the blank space here between the bracket and "and" (https://www.consul.io/intro/index.html)  and the [in-depth architecture